# Preprocessing

In [126]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [127]:
df_cc = pd.read_csv('/content/CC GENERAL.csv')

In [128]:
df_cc.head()

CUST_ID      BALANCE  ...  PRC_FULL_PAYMENT  TENURE
0  C10001    40.900749  ...          0.000000      12
1  C10002  3202.467416  ...          0.222222      12
2  C10003  2495.148862  ...          0.000000      12
3  C10004  1666.670542  ...          0.000000      12
4  C10005   817.714335  ...          0.000000      12

[5 rows x 18 columns]

CUSTID : Identification of Credit Card holder (Categorical)  
BALANCE : Balance amount left in their account to make purchases (  
BALANCEFREQUENCY : How frequently the Balance is updated, score between 0 and 1 (1 = frequently updated, 0 = not frequently updated)  
PURCHASES : Amount of purchases made from account  
ONEOFFPURCHASES : Maximum purchase amount done in one-go  
INSTALLMENTSPURCHASES : Amount of purchase done in installment  
CASHADVANCE : Cash in advance given by the user  
PURCHASESFREQUENCY : How frequently the Purchases are being made, score between 0 and 1 (1 = frequently purchased, 0 = not frequently purchased)  
ONEOFFPURCHASESFREQUENCY : How frequently Purchases are happening in one-go (1 = frequently purchased, 0 = not frequently purchased)  
PURCHASESINSTALLMENTSFREQUENCY : How frequently purchases in installments are being done (1 = frequently done, 0 = not frequently done)  
CASHADVANCEFREQUENCY : How frequently the cash in advance being paid  
CASHADVANCETRX : Number of Transactions made with "Cash in Advanced"  
PURCHASESTRX : Numbe of purchase transactions made  
CREDITLIMIT : Limit of Credit Card for user  
PAYMENTS : Amount of Payment done by user  
MINIMUM_PAYMENTS : Minimum amount of payments made by user  
PRCFULLPAYMENT : Percent of full payment paid by user  
TENURE : Tenure of credit card service for user  

In [129]:
df_cc.head()

CUST_ID      BALANCE  ...  PRC_FULL_PAYMENT  TENURE
0  C10001    40.900749  ...          0.000000      12
1  C10002  3202.467416  ...          0.222222      12
2  C10003  2495.148862  ...          0.000000      12
3  C10004  1666.670542  ...          0.000000      12
4  C10005   817.714335  ...          0.000000      12

[5 rows x 18 columns]

In [130]:
df_cc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8950 entries, 0 to 8949
Data columns (total 18 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   CUST_ID                           8950 non-null   object 
 1   BALANCE                           8950 non-null   float64
 2   BALANCE_FREQUENCY                 8950 non-null   float64
 3   PURCHASES                         8950 non-null   float64
 4   ONEOFF_PURCHASES                  8950 non-null   float64
 5   INSTALLMENTS_PURCHASES            8950 non-null   float64
 6   CASH_ADVANCE                      8950 non-null   float64
 7   PURCHASES_FREQUENCY               8950 non-null   float64
 8   ONEOFF_PURCHASES_FREQUENCY        8950 non-null   float64
 9   PURCHASES_INSTALLMENTS_FREQUENCY  8950 non-null   float64
 10  CASH_ADVANCE_FREQUENCY            8950 non-null   float64
 11  CASH_ADVANCE_TRX                  8950 non-null   int64  
 12  PURCHA

In [131]:
# menghilangkan kolom id (tidak diperlukan untuk clustering)
df_cc.drop('CUST_ID', axis=1, inplace=True)

In [132]:
# menghitung banyaknya missing value di setiap kolom
df_cc.isnull().sum()

BALANCE                               0
BALANCE_FREQUENCY                     0
PURCHASES                             0
ONEOFF_PURCHASES                      0
INSTALLMENTS_PURCHASES                0
CASH_ADVANCE                          0
PURCHASES_FREQUENCY                   0
ONEOFF_PURCHASES_FREQUENCY            0
PURCHASES_INSTALLMENTS_FREQUENCY      0
CASH_ADVANCE_FREQUENCY                0
CASH_ADVANCE_TRX                      0
PURCHASES_TRX                         0
CREDIT_LIMIT                          1
PAYMENTS                              0
MINIMUM_PAYMENTS                    313
PRC_FULL_PAYMENT                      0
TENURE                                0
dtype: int64

Pada hasil output di atas di temukan terdapat,  
1 missing value pada kolom CREDIT_LIMIT,  
313 missing value pada kolom MINIMUM_PAYMENT  

In [133]:
df_cc.describe()

BALANCE  BALANCE_FREQUENCY  ...  PRC_FULL_PAYMENT       TENURE
count   8950.000000        8950.000000  ...       8950.000000  8950.000000
mean    1564.474828           0.877271  ...          0.153715    11.517318
std     2081.531879           0.236904  ...          0.292499     1.338331
min        0.000000           0.000000  ...          0.000000     6.000000
25%      128.281915           0.888889  ...          0.000000    12.000000
50%      873.385231           1.000000  ...          0.000000    12.000000
75%     2054.140036           1.000000  ...          0.142857    12.000000
max    19043.138560           1.000000  ...          1.000000    12.000000

[8 rows x 17 columns]

In [134]:
# menghilangkan 1 baris yang memiliki missing value pada kolom CREDIT LIMIT
to_drop = df_cc.loc[df_cc.isnull()['CREDIT_LIMIT']].index
df_cc.drop(to_drop, axis=0, inplace=True)

In [135]:
df_cc.head()
df_cc.shape

(8949, 17)

In [136]:
# melakukan imputasi pada kolom missing value dengan nilai mean
df_cc.fillna(df_cc.mean(), inplace=True)

In [137]:
import seaborn as sns

In [138]:
from scipy import stats

In [139]:
# mencari nilai interquantile
Q1 = df_cc.quantile(0.25)
Q3 = df_cc.quantile(0.75)
IQR = Q3 - Q1
print(IQR)

BALANCE                             1926.007066
BALANCE_FREQUENCY                      0.111111
PURCHASES                           1070.370000
ONEOFF_PURCHASES                     577.830000
INSTALLMENTS_PURCHASES               468.650000
CASH_ADVANCE                        1113.868654
PURCHASES_FREQUENCY                    0.833334
ONEOFF_PURCHASES_FREQUENCY             0.300000
PURCHASES_INSTALLMENTS_FREQUENCY       0.750000
CASH_ADVANCE_FREQUENCY                 0.222222
CASH_ADVANCE_TRX                       4.000000
PURCHASES_TRX                         16.000000
CREDIT_LIMIT                        4900.000000
PAYMENTS                            1517.996470
MINIMUM_PAYMENTS                     693.429330
PRC_FULL_PAYMENT                       0.142857
TENURE                                 0.000000
dtype: float64


In [140]:
# mencari data outlier menggunakan metode Interquantile
(df_cc < (Q1 - 1.5 * IQR)) | (df_cc > (Q3 + 1.5 * IQR))

BALANCE  BALANCE_FREQUENCY  ...  PRC_FULL_PAYMENT  TENURE
0       False              False  ...             False   False
1       False              False  ...             False   False
2       False              False  ...             False   False
3       False               True  ...             False   False
4       False              False  ...             False   False
...       ...                ...  ...               ...     ...
8945    False              False  ...              True    True
8946    False              False  ...             False    True
8947    False              False  ...             False    True
8948    False              False  ...             False    True
8949    False               True  ...             False    True

[8949 rows x 17 columns]

In [141]:
# menghilangkan outlier
df_cc_out = df_cc[~((df_cc < (Q1 - 1.5 * IQR)) |(df_cc > (Q3 + 1.5 * IQR))).any(axis=1)]
df_cc_out.shape

(3008, 17)

In [142]:
from sklearn.preprocessing import StandardScaler

In [143]:
scaler = StandardScaler()
std = scaler.fit_transform(df_cc_out)
df_cc_norm = pd.DataFrame(std, columns=df_cc_out.columns)
df_cc_norm.head()

BALANCE  BALANCE_FREQUENCY  ...  PRC_FULL_PAYMENT  TENURE
0 -1.131269          -2.612474  ...         -0.413311     0.0
1 -0.380238           0.338454  ...         -0.413311     0.0
2  0.592313           0.338454  ...         -0.413311     0.0
3 -0.189572           0.338454  ...         -0.413311     0.0
4  1.509931           0.338454  ...         -0.413311     0.0

[5 rows x 17 columns]